File: app/src/main/java/forpdateam/ru/forpda/model/repository/ProfileRepository.java
package forpdateam.ru.forpda.model.repository;

import forpdateam.ru.forpda.api.profile.Profile;
import forpdateam.ru.forpda.api.profile.models.ProfileModel;
import forpdateam.ru.forpda.model.system.SchedulersProvider;
import io.reactivex.Observable;

/**
 * Created by radiationx on 02.01.18.
 */

public class ProfileRepository {

    private SchedulersProvider schedulers;
    private Profile profileApi;

    public ProfileRepository(SchedulersProvider schedulers, Profile profileApi) {
        this.schedulers = schedulers;
        this.profileApi = profileApi;
    }

    public Observable<ProfileModel> loadProfile(String url) {
        return Observable.fromCallable(() -> profileApi.getProfile(url))
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui());
    }

    public Observable<Boolean> saveNote(String note) {
        return Observable.fromCallable(() -> profileApi.saveNote(note))
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui());
    }
}

File: app/src/main/java/forpdateam/ru/forpda/model/repository/profile/ProfileRepository.kt
package forpdateam.ru.forpda.model.repository.profile

import forpdateam.ru.forpda.entity.remote.others.user.ForumUser
import forpdateam.ru.forpda.entity.remote.profile.ProfileModel
import forpdateam.ru.forpda.model.AuthHolder
import forpdateam.ru.forpda.model.data.cache.forumuser.ForumUsersCache
import forpdateam.ru.forpda.model.data.remote.api.profile.ProfileApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import ru.radiationx.coretypes.UserId
import javax.inject.Inject

/**
 * Created by radiationx on 02.01.18.
 */

class ProfileRepository @Inject constructor(
    private val profileApi: ProfileApi,
    private val authHolder: AuthHolder,
    private val forumUsersCache: ForumUsersCache
) {

    fun observeCurrentUser(): Flow<ForumUser?> {
        return authHolder.observe().flatMapLatest {
            if (it.isAuth()) {
                forumUsersCache.observeUserById(it.userId)
            } else {
                flowOf(null)
            }
        }
    }

    suspend fun getCurrentUser(): ForumUser? {
        val authState = authHolder.get()
        if (!authState.isAuth()) {
            return null
        }
        val cachedUser = forumUsersCache.getUserById(authState.userId)
        if (cachedUser != null) {
            return cachedUser
        }
        return loadSelf().user
    }

    suspend fun loadSelf(): ProfileModel {
        val authState = authHolder.get()
        require(authState.isAuth()) {
            "Invalid auth state $authState"
        }
        return loadProfile(authState.userId)
    }

    suspend fun loadProfile(userId: UserId): ProfileModel {
        return profileApi.getProfile(userId).also {
            forumUsersCache.saveUser(it.user)
        }
    }

    suspend fun saveNote(note: String): Boolean {
        return profileApi.saveNote(note)
    }
}
